feat: v0 and v1 encodings

eventualbuddha opened the pull request:
There's a fair amount of work to do to clean this up and document it, but the basics are here. Check out the tests for [v0](https://github.com/votingworks/ballot-encoder/blob/v0-and-v1/src/v0/index.test.ts) and [v1](https://github.com/votingworks/ballot-encoder/blob/v0-and-v1/src/v1/index.test.ts) to get a sense of the structure.

eventualbuddha committed:
initial commit

eventualbuddha committed:
chore: require node >= v12

eventualbuddha committed:
chore: update sample election data and export it

eventualbuddha committed:
refactor: minor clarifications to v0 encoding API

eventualbuddha committed:
feat: finish first working version of v1 encoding

benadida reviewed (approved):
Brian, this looks awesome. Very meticulous. I imagine you want ot make it even better, but the tests are incredibly thorough and make me happy. I recommend moving forward as soon as you're game.

eventualbuddha committed:
refactor: use the `vote` helper for the v0 tests

eventualbuddha committed:
fix: include contest id in decoding errors

eventualbuddha committed:
test: add a round-trip test for the main v1 API

eventualbuddha committed:
feat: allow writing sequences of uint8 in one call

eventualbuddha committed:
feat: allow skipping uints while reading

This is going to be used to skip the prelude for used to detect format & version.

eventualbuddha committed:
feat: copy over `getPartyPrimaryAdjectiveFromBallotStyle` from bmd

Also adds some documentation.

eventualbuddha committed:
docs: describe v0 as a BNF

Includes a few minor fixes.

eventualbuddha committed:
feat: add `VX` + version marker to v1 format

Adds `detect` functions for both v0 and v1.

eventualbuddha committed:
feat: add library-level encode/decode/detect functions

These delegate to the individual encoder versions as needed.

eventualbuddha committed:
docs: add a root-level README

eventualbuddha committed:
refactor: rename `Ballot` to `CompletedBallot`
